core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java

```diff
@@ -4107,6 +4113,97 @@ public static List arrayExcept(List list1, List list2) {
     return new ArrayList<>(result);
   }
 
+  /** Support the ARRAY_INSERT(array, pos, val) function. */
+  public static @Nullable List arrayInsert(List baselist, Object pos, Object val) {
+    if (baselist == null || pos == null) {
+      return null;
+    }
+    int posInt = (int) pos;
+    if (posInt == 0) {
+      throw new IllegalArgumentException("The index 0 is invalid. "
+          + "An index shall be either < 0 or > 0 (the first element has index 1)");
+    }
+
+    boolean usePositivePos = posInt > 0;
+
+    if (usePositivePos) {
+      int newArrayLength = Math.max(baselist.size() + 1, posInt);
+
+      if (newArrayLength > MAX_ARRAY_LENGTH) {
+        throw new IndexOutOfBoundsException(
+            String.format(Locale.ROOT, "The new array length %s exceeds the allowed limit.",
+                newArrayLength));
+      }
+
+      Object[] newArray = new Object[newArrayLength];
+
+      int posIndex = posInt - 1;
```

Comment:
Can we use `list.add(pos,element)` ?

Reply:
@NobiGo I'm afraid not. 

because our logic here is to move the element on the right of the insertion point to the right, leave the left side unchanged, leaving a space for the target element. If we use `list.add` to insert the left element, there is no problem. When inserting the right element, because we want to reserve a position, it will go out of bounds at this time.

A simple case can explain:
List<String> list = new ArrayList<>();
list.add(0, "A");
list.add(1, "A"); 
list.add(3, "A"); // exception